cpp/src/arrow/extension_type_test.cc

```diff
@@ -333,4 +334,144 @@ TEST_F(TestExtensionType, ValidateExtensionArray) {
   ASSERT_OK(ext_arr4->ValidateFull());
 }
 
+class TensorArray : public ExtensionArray {
+ public:
+  using ExtensionArray::ExtensionArray;
+};
+
+class TensorArrayType : public ExtensionType {
+ public:
+  explicit TensorArrayType(const std::shared_ptr<DataType>& type,
+                           const std::vector<int64_t>& shape,
+                           const std::vector<int64_t>& strides)
+      : ExtensionType(type), type_(type), shape_(shape), strides_(strides) {}
+
+  std::shared_ptr<DataType> type() const { return type_; }
+  std::vector<int64_t> shape() const { return shape_; }
+  std::vector<int64_t> strides() const { return strides_; }
+
+  std::string extension_name() const override {
+    std::stringstream s;
+    s << "ext-array-tensor-type<type=" << *storage_type() << ", shape=(";
+    for (uint64_t i = 0; i < shape_.size(); i++) {
+      s << shape_[i];
+      if (i < shape_.size() - 1) {
+        s << ", ";
+      }
+    }
+    s << "), strides=(";
+    for (uint64_t i = 0; i < strides_.size(); i++) {
+      s << strides_[i];
+      if (i < strides_.size() - 1) {
+        s << ", ";
+      }
+    }
+    s << ")>";
+    return s.str();
+  }
+
+  bool ExtensionEquals(const ExtensionType& other) const override {
+    return this->shape() == static_cast<const TensorArrayType&>(other).shape();
```

Comment:
I think this should be:

```suggestion
    return this->shape().size() == static_cast<const TensorArrayType&>(other).shape().size();
```

Reply:
Ah, I misunderstood your suggestion. Why would you need it so loose?

Reply:
Added a `ndim` method and used it for equality comparison.

Reply:
> Ah, I misunderstood your suggestion. Why would you need it so loose?

More practically, this means that one cannot have for e.g. a `(10, 5, 4)` shape Tensor and a `(10, 6, 2)` Tensor in separate parquet files in the same dataset -- IIRC the Dataset API will complain that their metadata doesn't agree (due to the strict equality comparison). I do run into these sort of cases in the datasets that I deal with.

I would argue that parameterising Tensors Types on `shape` and `stride` introduces an infinite number of parameterisations and I'm not sure that this class of parameterisations is useful. This doesn't imply that `shape` and `stride` should not be attributes on a Tensor Type!

Reply:
> > Ah, I misunderstood your suggestion. Why would you need it so loose?
> 
> Practically speaking, this means that one cannot have for e.g. a `(10, 5, 4)` shape Tensor and a `(10, 6, 2)` Tensor in separate parquet files in the same dataset -- IIRC the Dataset API will complain that their metadata doesn't agree (due to the strict equality comparison). I do run into these sort of cases in the datasets that I deal with.
> 
> More formally, I would argue that parameterising Tensors Types on `shape` and `stride` introduces an infinite number of parameterisations and I'm not sure that this class of parameterisations is useful. This doesn't imply that `shape` and `stride` should not be attributes on a Tensor Type!

Got it. Infinite types sounds a bit redundant indeed.
The only reason I can think of to have dimensions and strides in equality comparison then is if we did some compute kernels that needed to identify type in advance. But even that can probably be solved at runtime.

In that case do we even want to keep `ndim` for equality comparison?

Reply:
> In that case do we even want to keep ndim for equality comparison?

This is a good question. I might lean towards saying not, but I'm not a maintainer. I guess it depends on how the type is used throughout the rest of the Arrow ecosystem -- you mentioned the Compute Engine for example.

For instance, Numba parameterise their [Array](https://github.com/numba/numba/blob/e617b39a0b4b23d7b69d16f482fd66b4ac6cc307/numba/core/types/npytypes.py#L423) type on `dtype`, `ndim`, `layout`, `readonly` and `alignment`, I would guess because these properties are useful for generating efficient LLVM code.

I'm not suggesting that Arrow should copy Numba's Type parameterisation, but its similar enough to provide food for thought.

Reply:
Also, I think that Google jax include shape in their Type Parameterisations, AFAIK because the XLA compiler can produce very optimal code when array bounds are known. But it does lead to these kind of sharp edges:

- Search for "functions with argument-value dependent shapes" under https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#structured-control-flow-primitives
- https://github.com/google/jax/issues/5100#issuecomment-738904018

Reply:
I removed the `ndim` check and made equality comparison only check for type name (`ext-array-tensor-type`).

> > In that case do we even want to keep ndim for equality comparison?
>
> This is a good question. I might lean towards saying not, but I'm not a maintainer. I guess it depends on how the type is used throughout the rest of the Arrow ecosystem -- you mentioned the Compute Engine for example.

I'm not aware of active work on tensor computation in Arrow at the moment so I don't think there were any decisions made on this yet. It is super interesting to see what were the trade-offs in other places (numba, jax) though.

At the moment this is an `ExtensionArray` in `cpp/src/arrow/extension_type_test.cc`. My understanding is there are no plans to make these extension arrays available elsewhere @jorisvandenbossche @wesm ?

Reply:
> I removed the `ndim` check and made equality comparison only check for type name (`ext-array-tensor-type`).

I don't really understand the argument about not being strict in the type equality. 

Currently, the example code here uses `FixedSizeListArray` as the underlying storage type, for _constant_ dimension (and I would say constant shape and strides, as well?) for the full array. 

So if you have multiple files in the same dataset where the shape of the tensors differs between files, how can that ever be combined in a single schema / table? FixedSizeListArray with a different size are not compatible with each other. 

If you want to be flexible regarding the shape of each tensor value, it seems that is rather what is covered in https://issues.apache.org/jira/browse/ARROW-8714, where the storage type could be a struct with one field the data as a variable sized list array and the other field storing the shape of each tensor value.

Reply:
Hm, I would agree with Joris here. I'll reintroduce parameters into the name.